### PR TITLE
CI: Check for whitespaces before class

### PR DESCRIPTION
**Pull request by Ayowolet**
xref #28209, https://github.com/python-sprints/pandas-mentoring/issues/161

Validate that we use
```python
class Foo:
    """
    Doc.
    """
    pass
```
instead of:
```python
class Bar:

    """
    Doc.
    """
    pass
```
(note the blank line between `class  Bar:` and the docstring.

### CODE REVIEW

**Comment by Ayowolet** [comment]
#28209 @datapythonista I have created the pull request, was a little unsure but does it look okay?

File: ci/code_checks.sh
Comment (datapythonista) [inline ci/code_checks.sh]: I don't think we need this comment, it's already explained in the message in the next line.

File: ci/code_checks.sh
Comment (datapythonista) [inline ci/code_checks.sh]:
```suggestion
    invgrep -R --include="*.py" --include="*.pyx" -E 'class.*:\n\n( )+"""' .
```

The goal here is to check all the Python files in the repo, to standardize classes. The last path `doc/source/` is checking only in the documentation directory, where I don't think there are python files, so we want to look in the root instead

**Review by datapythonista — COMMENTED** [review]
Great job @Ayowolet, looks almost perfect, just added couple of comments.

Also, would be great if you can find any class definition in the code, and add the blank line we're trying to detect here. This way, the continuous integration will become red, and we'll see that the check you're adding is working. After we see that it's working, you'll have to revert the blank line, before we can merge.

cc @jbrockmendel

**Comment by jbrockmendel** [comment]
pending @datapythonista's suggestion about adding-then-reverting a failing case: LGTM

**Comment by Ayowolet** [comment]
@datapythonista should I add the change to the class for the test to this PR?

**Comment by datapythonista** [comment]
Exactly.

The idea is that you add a blank line after the `class Whatever:` line of any class in our code, and push it. Then we wait for the CI, and if your changes are working, it should be red and report the error. If that happens, you just delete the blank line again, and we merge.

**Comment by datapythonista** [comment]
Looks like your changes are not working. If you go the the CI, in the `Checks` build, and the step `Looking for unwanted patterns`, you'll see some errors.

What I'd do is next:
- Just leave one blank line and not two in what you added last, and remove any space in those blank lines
- Execute locally `./ci/code_checks.sh patterns`
- Fix the problems in your expression until it reports that file
- Then commit and push to this branch

Let me know if you need help. Thanks!

**Comment by jbrockmendel** [comment]
@Ayowolet can you rebase

**Comment by Ayowolet** [comment]
@jbrockmendel, I'm not sure I understand what you mean by rebase...

**Comment by jbrockmendel** [comment]
> I'm not sure I understand what you mean by rebase...

Near the bottom of this page there is a gray warning saying "This branch has conflicts that must be resolved".  To resolve these, you'll need to run `git pull upstream master` and fix any merge conflicts that show up.

**Comment by Ayowolet** [comment]
okay, thank you I will do that

**Comment by jbrockmendel** [comment]
@Ayowolet can you rebase and see why the check is failing on azure

**Comment by Ayowolet** [comment]
If you can please help with the steps for this.. I tried the other method
you suggested but I think the internet connection was too slow and the page
just kept loading for days...

**Comment by TomAugspurger** [comment]
@Ayowolet the logs have been removed now, but here are a few failures.

```
 ./pandas/tests/test_algos.py#L37

./pandas/tests/test_algos.py(37,1): error W293: blank line contains whitespace
```

```
 ./pandas/tests/test_algos.py#L38

./pandas/tests/test_algos.py(38,5): error E303: too many blank lines (2)

```

```
 Build log #L42

Bash exited with code '2'.

```

I'm not sure what the last one means. You might try running `./ci/code_checks.sh lint` locally.

**Comment by jbrockmendel** [comment]
> I tried the other method you suggested

`git pull upstream master` didn't work?  That should be a fairly light-weight request, so I don't know what would cause that.

The azure page having trouble loading might make more sense, so I'll check that.  Looks like the linting is failing because of excess whitespace on tests.test_algos lines 37 and 38

**Comment by WillAyd** [comment]
I think this is OK. @Ayowolet can you fix the class docstring again so CI turns green?

**Comment by jbrockmendel** [comment]
@Ayowolet did the trouble with rebasing get resolved?

**Review by datapythonista — APPROVED** [review]
lgtm

I think the regex could be improved, but this is surely good enough for now, and we can improve it if we ever get false positives.

**Comment by jreback** [comment]
thanks @Ayowolet

**Pull request merged**
